mesh/mesh.cpp

```diff
@@ -7094,6 +7095,59 @@ bool Mesh::NonconformingDerefinement(Array<double> &elem_error,
    return true;
 }
 
+bool Mesh::GeneralDerefinement(Array<int>& derefs)
+{
+   MFEM_VERIFY(ncmesh, "Only supported for non-conforming meshes.");
+   MFEM_VERIFY(!NURBSext, "Derefinement of NURBS meshes is not supported. "
+               "Project the NURBS to Nodes first.");
+
+   DeleteLazyTables();
+
+   if (!derefs.Size()) { return false; }
+
+   const Table &dt = ncmesh->GetDerefinementTable();
+
+   // map elements to rows in table
+   Table dtT;
+   Transpose(dt, dtT);
+
+   // Find the set of rows in the refinement table which correspond to
+   // the specified elements to derefine
+   std::set<int> set_rows;
+   for (int i = 0; i < derefs.Size(); i++) {
+      int el = derefs[i];
+      const int *row = dtT.GetRow(el);
+      set_rows.insert(row[0]);
+   }
+
+   Array<int> deref_rows;
+   std::set<int>::iterator it;
+   for (it = set_rows.begin(); it != set_rows.end(); ++it) {
```

Comment:
I had the same thoughts regarding the API, which is why I thought it was worth discussing before going any further with it.  Supplying the coarsening table to the user is one possibility.

Maybe another possibility is to use a callback mechanism, where the Mesh class internally iterates over the derefinement table, and calls the callback to make a derefinement decision?  Does that address the dependency issue you mention above, or is it more subtle than that?

Reply:
I'm starting to like the "callback" idea. A simple inline function could be defined that goes through the derefinement table and calls a user supplied functor that checks whether a derefinement should be performed or not. Then it would call the usual internal stuff.

Reply:
I'll give a shot at prototyping that idea.